File: python3/find-the-minimum-amount-of-time-to-brew-potions.py
class Solution:
    def minTime(self, skill: list[int], mana: list[int]) -> int:
        wizard_ready = [0] * len(skill)
        for mana_required in mana:
            potion_ready = wizard_ready[0]

            for wizard, s in enumerate(skill):
                potion_ready = max(potion_ready, wizard_ready[wizard])
                potion_ready += s * mana_required
                wizard_ready[wizard] = potion_ready

            # adjust actual time when each wizard finished with this potion
            for i in range(len(skill) - 2, -1, -1):
                wizard_ready[i] = wizard_ready[i + 1] - mana_required * skill[i + 1]
        return wizard_ready[-1]

File: go/find-the-minimum-amount-of-time-to-brew-potions.go
func minTime(skill []int, mana []int) int64 {
	var wizardReady []int64 = make([]int64, len(skill))
	for _, manaRequired := range mana {
		var potionReady int64 = wizardReady[0]

		for wizard, s := range skill {
			potionReady = max(potionReady, wizardReady[wizard])
			potionReady += int64(manaRequired) * int64(s)
			wizardReady[wizard] = potionReady
		}

		for i := len(skill) - 2; i >= 0; i-- {
			wizardReady[i] = wizardReady[i + 1] - int64(manaRequired) * int64(skill[i + 1])
		}
	}
	return wizardReady[len(wizardReady) - 1]
}
